include/pybind11/stl_bind.h

```diff
@@ -785,7 +785,7 @@ class_<Map, holder_type> bind_map(handle scope, const std::string &name, Args &&
         [](Map &m, const KeyType &k) -> MappedType & {
             auto it = m.find(k);
             if (it == m.end()) {
-                throw key_error();
+                throw key_error(str(cast(k)).cast<std::string>());
```

Comment:
Looking at this more closely, there are a bunch of things that can go wrong:

* `cast(k)` may fail,
* `str(...)` may fail,
* `.cast<std::string>()` may fail.

Any of those failures will be a secondary error, masking the desired `KeyError`. This can be extremely confusing.

Making this robust is not easy.

What's needed is something like (untested; only to give the idea):

```
str key_as_str;
try {
  key_as_str = str(cast(k));
}
catch (const std::exception&) {
    throw key_error();
}
set_error(PyExc_KeyError, key_as_tr);
throw error_already_set(); 
```

This would be implemented in a helper function.

Another though: if str() fails, try repr().

Oh, there should also be a size limit. I.e. you don't want to produce a 1M numpy array as a str.

Sorry to make this difficult, but these things sure will bite, someone.

Reply:
@rwgk PTAL. I added a test for long keys. I don't know what kind of object would fail in str of cast - happy to add a test if you have a suggestion. What is the benefit of `set_error + throw error_already_set()` over `throw key_error()` ?